### PR TITLE
[ospec] Improve the dupe detection code

### PR DESCRIPTION
**Pull request by pygy**
It will run all tests, even if more than two share the same name. This covers the case were the attempted deduped name already exists.

AFAIK mutating a function argument can cause deopts in V8. Removing these mutations doesn't seem to affect the run time but better be safe than sorry...

At last `[\w_]` is redundant since `\w` matches `_`.

### CODE REVIEW

**Comment by isiahmeadows** [comment]
@pygy 

> AFAIK mutating a function argument can cause deopts in V8. Removing these mutations doesn't seem to affect the run time but better be safe than sorry...

[Only if you're in sloppy mode *and* referencing `arguments`.](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#31-reassigning-a-defined-parameter-while-also-mentioning-arguments-in-the-body-in-sloppy-mode-only-typical-example) Not the case here, so it'd be surprising if you *did* see any significant performance difference.

File: ospec/ospec.js [hunk 1]
```diff
@@ -112,12 +110,15 @@ module.exports = new function init() {
 			}
 		}
 	}
-	function unique(subject, predicate) {
-		if (ctx.hasOwnProperty(subject) && predicate) {
-			console.warn("A test named `" + subject + "` was already defined")
-			subject = subject + "*"
+	var dedupeCounter = 0
+	function unique(subject) {
+		var res = subject
```
Comment (isiahmeadows) [inline ospec/ospec.js]: Just reuse `subject`. My comment explains why.

File: ospec/ospec.js [hunk 2]
```diff
@@ -112,12 +110,15 @@ module.exports = new function init() {
...
+		if (ctx.hasOwnProperty(subject)) {
```
Comment (isiahmeadows) [inline ospec/ospec.js]: Use `Object.prototype.hasOwnProperty` directly here, so it doesn't get screwed up over a test literally named `hasOwnProperty`.

Comment (pygy) [inline ospec/ospec.js]: Good catch!

**Comment by pygy** [comment]
@isiahmeadows thanks for the review, the patch is updated. I've fixed another `hasOwnProperty` bug.

Thanks also for refreshing my Chrome deopt knowledge... Indeed there were no slow downs.

Last time I checked, mutating arguments was slowing down Safari, though, so it should be kept in mind for code primarily intended for browsers.

**Pull request merged**